common/events_common.h

```diff
@@ -0,0 +1,356 @@
+
+
+/* helper API to print variable type */
+/*
+ *Usage:
+ *    const int ci = 0;
+ *    std::cout << type_name<decltype(ci)>() << '\n';
+ *
+ *    map<string, string> l;
+ *    std::cout << type_name<decltype(l)>() << '\n';
+ *
+ *    tt_t t;
+ *    std::cout << type_name<decltype(t)>() << '\n';
+ *    std::cout << type_name<decltype(tt_t)>() << '\n';
+ */
+template <typename T> std::string type_name();
+
+template <class T>
+std::string
+type_name()
+{
+    typedef typename std::remove_reference<T>::type TR;
+    std::unique_ptr<char, void(*)(void*)> own
+        (
+            abi::__cxa_demangle(typeid(TR).name(), nullptr,
+                nullptr, nullptr),
+            std::free
+        );
+    std::string r = own != nullptr ? own.get() : typeid(TR).name();
+    if (std::is_const<TR>::value)
+        r += " const";
+    if (std::is_volatile<TR>::value)
+        r += " volatile";
+    if (std::is_lvalue_reference<T>::value)
+        r += "&";
+    else if (std::is_rvalue_reference<T>::value)
+        r += "&&";
+    return r;
+}
+
+
+template <class T>
+std::string
+get_typename(T &val)
+{
+    return type_name<decltype(val)>();
+}
+
+
+/* map to human readable str; Useful for error reporting. */
+template <typename Map>
+string
+map_to_str(const Map &m)
+{
+    stringstream _ss;
+    _ss << "{";
+    for (const auto elem: m) {
+        _ss << "{" << elem.first << "," << elem.second << "}";
+    }
+    _ss << "}";
+    return _ss.str();
+}
+
+// Some simple definitions
+//
+typedef map<string, string> map_str_str_t;
+
+/*
+ * Config that can be read from init_cfg
+ */
+#define INIT_CFG_PATH "/etc/sonic/init_cfg.json"
+#define CFG_EVENTS_KEY "events"
+
+/* configurable entities' keys */
+/* zmq proxy's xsub & xpub end points */
+#define XSUB_END_KEY "xsub_path"       
+#define XPUB_END_KEY "xpub_path"
+
+/* Eventd service end point; All service req/resp occur via this path */
+#define REQ_REP_END_KEY "req_rep_path"
+
+/* Internal proxy to service path for capturing events for caching */
+#define CAPTURE_END_KEY "capture_path"
+
+#define STATS_UPD_SECS "stats_upd_secs"
+#define CACHE_MAX_CNT "cache_max_cnt"
+
+/* init config from file */
+void read_init_config(const char *fname);
+
+/* Read config entry for a key */
+string get_config(const string key);
+
+template<typename T>
+T get_config_data(const string key, T def)
+{
+    string s(get_config(key));
+
+    if (s.empty()) {
+        return def;
+    }
+    else {
+        T v;
+        stringstream ss(s);
+
+        ss >> v;
+
+        return v;
+    }
+}
+
+
+const string get_timestamp();
+
+/*
+ * Way to serialize map or vector
+ * boost::archive::text_oarchive could be used to archive any struct/class
+ * but that class needs some additional support, that declares
+ * boost::serialization::access as private friend and couple more tweaks.
+ * The std::map & vector inherently supports serialization.
+ */
+template <typename Map>
+int
+serialize(const Map& data, string &s)
```

Comment:
Why does serialize require address of Map?

Reply:
Just to clarify my question, any need for passing by reference here?

Reply:
Reference is same as pointer, but assures that the pointer can't be NULL.
So where we mandatorily expect data, we could use reference and compiler takes care of it.